src/lib.rs

```diff
@@ -3799,46 +3795,96 @@ safety_comment! {
     unsafe_impl_for_power_set!(A, B, C, D, E, F, G, H, I, J, K, L -> M => Immutable for opt_extern_c_fn!(...));
 }
 
-macro_rules! impl_traits_for_atomics {
-    ($($atomics:ident [$inners:ident]),* $(,)?) => {
-        $(
-            impl_for_transparent_wrapper!(TryFromBytes for $atomics [UnsafeCell<$inners>]);
-            impl_for_transparent_wrapper!(FromZeros for $atomics [UnsafeCell<$inners>]);
-            impl_for_transparent_wrapper!(FromBytes for $atomics [UnsafeCell<$inners>]);
-            impl_for_transparent_wrapper!(IntoBytes for $atomics [UnsafeCell<$inners>]);
-        )*
-    };
+/// Implements the zerocopy traits for atomic integer types.
+///
+/// Note that we only implement traits for atomics types if the corresponding
+/// [`#[cfg(target_has_atomic)]`][target_has_atomic] configuration is true.
+///
+/// Note: On targets like `thumbv6m-none-eabi` (which has an [`AtomicU32`]
+/// type but `#[cfg(target_has_atomic = "32")]` is false), we don't implement
+/// the zerocopy traits. We will be able to handle this when Rust stabilizes
+/// [`#[cfg(target_has_atomic_load_store)]`][target_has_atomic_load_store].
+///
+/// [target_has_atomic]: https://doc.rust-lang.org/reference/conditional-compilation.html#target_has_atomic
+/// [target_has_atomic_load_store]: https://www.github.com/rust-lang/rust/issues/94039
+///
+/// # Safety
+///
+/// `$atomic` must be an atomic type with corresponding native type `$native`.
+#[cfg(zerocopy_atomics)]
+macro_rules! unsafe_impl_traits_for_atomics {
+    ($($atomic:ident [$native:ident]),*) => {$(
+        // SAFETY: `$atomic` is an atomic type with native type `$native`.
+        unsafe impl util::Atomic for $atomic { type Native = $native; }
+
+        impl_for_transparent_wrapper!(TryFromBytes for $atomic);
+        impl_for_transparent_wrapper!(FromZeros for $atomic);
+        unsafe_impl_traits_for_atomics!(@from_bytes $atomic);
+        impl_for_transparent_wrapper!(IntoBytes for $atomic);
+        impl_known_layout!($atomic);
+    )*};
+    (@from_bytes AtomicBool) => {};
+    (@from_bytes $atomic:ty) => { impl_for_transparent_wrapper!(FromBytes for $atomic); };
 }
 
-#[rustfmt::skip]
-impl_traits_for_atomics!(
-    AtomicBool [bool],
-    AtomicI16 [i16], AtomicI32 [i32], AtomicI8 [i8], AtomicIsize [isize],
-    AtomicU16 [u16], AtomicU32 [u32], AtomicU8 [u8], AtomicUsize [usize],
-);
+#[cfg(zerocopy_atomics)]
+#[cfg(target_has_atomic = "8")]
+#[cfg_attr(doc_cfg, doc(cfg(target_has_atomic = "8")))]
+mod atomic_8 {
+    use super::*;
+    unsafe_impl_traits_for_atomics!(AtomicBool[bool], AtomicU8[u8], AtomicI8[i8]);
+    safety_comment! {
+        /// SAFETY:
+        /// Per [1], `AtomicBool`, `AtomicU8`, and `AtomicI8` have an alignment of 1.
+        ///
+        /// [1] Per https://doc.rust-lang.org/nightly/core/sync/atomic/struct.AtomicBool.html:
+        ///
+        ///     This type has the same size, alignment, and bit validity as [the native type].
```

Comment:
```suggestion
        /// SAFETY:
        /// Per [1], `AtomicBool`, `AtomicU8`, and `AtomicI8` have an alignment of 1.
        ///
        /// [1] Per [2], [3], and [4]:
        ///
        ///     This type has the same size, alignment, and bit validity as [the native type].
        ///
        /// [2] https://doc.rust-lang.org/nightly/core/sync/atomic/struct.AtomicBool.html
        /// [3] https://doc.rust-lang.org/nightly/core/sync/atomic/struct.AtomicU8.html
        /// [4] https://doc.rust-lang.org/nightly/core/sync/atomic/struct.AtomicI8.html
```

I know we've been fast-and-loose with this in the past, but I'm trying to move us towards being more precise with our safety comments.